src/org/sosy_lab/common/log/BasicLogManager.java

```diff
@@ -414,7 +418,95 @@ private void log0(Level priority, StackTraceElement stackElement, String msg) {
     logger.log(record);
   }
 
-  private String buildMessageText(Object... args) {
+  /** {@inheritDoc} */
+  @Override
+  public void logUserException(Level priority, Throwable e, Object... args) {
+    checkNotNull(e);
+    checkNotNull(args);
+    @Var String additionalMessage = null;
+
+    if (wouldBeLogged(priority)) {
+      additionalMessage = buildAdditionalMessageText(args);
+      StringBuilder logMessage = buildUserExceptionLogMessage(priority, e, additionalMessage);
+      @Var StackTraceElement frame = locateStackTraceElement(e);
+
+      log0(priority, frame, logMessage.toString());
+    }
+    // Enforce lazy string building for DEBUG level
+    if (additionalMessage == null) {
+      logDebugException(e, args);
+    } else {
+      // but keep calculated results
+      logDebugException(e, additionalMessage);
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void logfUserException(Level priority, Throwable e, String format, Object... args) {
```

Comment:
`logUserException` and `logfUserException` look almost identical, basically only one line changes if I see correctly. Can this be refactored?

Reply:
I shied away from doing this, cause of the awkard logic structure you mentioned, too.
With the mentioned changes, this will be easier.